### 2.İlk olarak, Sıcaklık verilerini Kaggle'dan indirin.
Hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu doğrusal bir regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. Tahmin edilen katsayılar istatistiksel olarak anlamlı mıdır? Tahmini katsayılar önceki beklentileriniz doğrultusunda mı? Tahmin edilen katsayıları yorumlayınız. Hedef ve açıklayıcı değişkenler arasındaki ilişkiler nelerdir?

Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Katsayılar istatistiksel olarak anlamlı mıdır? Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [2]:
tem_df = pd.read_csv("weatherHistory.csv")
tem_df.head(10)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   
5  2006-04-01 05:00:00.000 +0200  Partly Cloudy        rain         9.222222   
6  2006-04-01 06:00:00.000 +0200  Partly Cloudy        rain         7.733333   
7  2006-04-01 07:00:00.000 +0200  Partly Cloudy        rain         8.772222   
8  2006-04-01 08:00:00.000 +0200  Partly Cloudy        rain        10.822222   
9  2006-04-01 09:00:00.000 +0200  Partly Cloudy        rain        13.772222   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   
5                  7.111111      0.85            13.9587   
6                  5.522222      0.95            12.3648   
7                  6.527778      0.89            14.1519   
8                 10.822222      0.82            11.3183   
9                 13.772222      0.72            12.5258   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   
5                   258.0          14.9569         0.0               1016.66   
6                   259.0           9.9820         0.0               1016.72   
7                   260.0           9.9820         0.0               1016.84   
8                   259.0           9.9820         0.0               1017.37   
9                   279.0           9.9820         0.0               1017.22   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.  
5  Partly cloudy throughout the day.  
6  Partly cloudy throughout the day.  
7  Partly cloudy throughout the day.  
8  Partly cloudy throughout the day.  
9  Partly cloudy throughout the day.

In [3]:
Y = tem_df["Apparent Temperature (C)"] - tem_df["Temperature (C)"]
X = tem_df[["Humidity","Wind Speed (km/h)"]]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 20 Apr 2021   Prob (F-statistic):               0.00
Time:                        19:35:42   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

hedef değişken ile özellikler arasındaki ilişki:

y = 2.43 - 3.02*humidity - 0.1193*Wind_speed

Tahmin edilen katsayılar pvalue < 0.05 olduğu için istatistiksel olarak anlamlıdır.

Hisssedilen sıcaklık ile nem arasında pozitif ilişki, rüzgar hızıyla da negatif ilişki olmasını beklerdim.Ancak nem katsayısı negatif ilişki olduğunu gösteriyor, tahminlerimle aynı doğrultuda değil.

Humidity miktarındaki bir birimlik artış sıcaklığı 3.02 birim azaltmakta, rüzgar hızındaki bir birimlik artış ise 0.1193 birim azaltmaktadır.

R2:humidity ve wind speed değişkenlerinin hedef değişkenle olan varyansını göstermektedir.Yani R2 sıcaklığın % 28.8 varyans oranını humidity ve wind speed değişken ile açıklayabilir.R2 ifadesinin her zaman daha yüksek olmasını isteriz bu değer sıcaklığı acıklamak için düşük bir değerdir.

Şimdi de modele nem ve rüzgar hızı etkileşimini ekleyelim.

In [4]:
tem_df["hum_wind_speed"] = tem_df["Humidity"]*tem_df["Wind Speed (km/h)"]

Y = tem_df["Apparent Temperature (C)"]-tem_df["Temperature (C)"] 
X = tem_df[["Humidity","Wind Speed (km/h)","hum_wind_speed"]]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 20 Apr 2021   Prob (F-statistic):               0.00
Time:                        19:35:42   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.0839      0.033      2.511      0.012       0.018       0.149
Humidity              0.1775      0.043      4.133      0.000       0.093       0.262
Wind Speed (km/h)     0.0905      0.002     36.797      0.000       0.086       0.095
hum_wind_speed       -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

hedef değişken ile özellikler arasındaki ilişki:

y = 0.0839 + 0.1775*humidity + 0.0905*Wind_speed - 0.2971*hum_wind_speed

Katsayılar istatistiksel olarak anlamlıdır.Bu kez humidity ve wind speed değişken ile sıcaklık arasında pozitif ilişki varken yeni olusturduğumuz humidity ve wind speed etkileşimli değişen ile negatif ilişki vardır. 

R2 değerinin ise artmış olduğunu görüyoruz.Bu kez sıcaklık hedef değişkeninin %34.1 lik varyans oranını ele aldığımız 3 değişkenle açıklayabildiğimizi ortaya koyar.

### 3. Ev fiyatları modeli
Bu alıştırmada, ev fiyatları modelinizi yorumlayacaksınız. Bu ödevi tamamlamak için, gist dosyasına, aşağıdaki görevlere ilişkin çözümlerinizi içeren Jupyter notebook'u bir link olarak ekleyin:

Ev Fiyatları verilerini Kaggle'dan indirin.
Ev fiyatları modelinizi tekrar çalıştırın ve sonuçları yorumlayın. Hangi özellikler istatistiksel olarak anlamlı ve hangileri değildir?
Şimdi, anlamsız özellikleri modelinizden hariç tutun. Bir şey değişti mi?
Ev fiyatları ile olan ilişkilerini belirleyerek istatistiksel olarak anlamlı katsayıları yorumlayın. Ev fiyatları üzerinde hangi özellikler daha belirgin bir etkiye sahiptir?
Sonuçlar size mantıklı geliyor mu? Değilse, potansiyel nedenleri açıklamaya çalışın.

In [2]:
pd.options.display.max_columns = None

house = pd.read_csv("train.csv")
house.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [3]:
KitchenQual_type = {"Po":0,"Fa":1,"TA":2,"Gd":3,"Ex":4}
house["KitchenQual_type"] = house["KitchenQual"].map(KitchenQual_type)

In [4]:
house2 = pd.get_dummies(house)

In [5]:
house2.columns[house2.isnull().any()]

Index(['LotFrontage', 'MasVnrArea', 'GarageYrBlt'], dtype='object')

In [6]:
house2.drop(['LotFrontage', 'MasVnrArea', 'GarageYrBlt'],axis=1,inplace=True)

### Model-1

Tüm değişkenleri modele aldığımızda cıkan sonuç: istatiksel olarak anlamsız olan birçok değişken mevcut fakat buna ragmen tüm değişkenler modelde yeraldığı için R2 değeri oldukça yüksek.

In [7]:
Y = house2["SalePrice"]
X = house2.drop(["SalePrice"],axis=1)

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.933
Model:                            OLS   Adj. R-squared:                  0.918
Method:                 Least Squares   F-statistic:                     66.50
Date:                Wed, 21 Apr 2021   Prob (F-statistic):               0.00
Time:                        06:58:27   Log-Likelihood:                -16576.
No. Observations:                1460   AIC:                         3.366e+04
Df Residuals:                    1208   BIC:                         3.499e+04
Df Model:                         251                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                  2.972e+04   1.35e+05      0.221      0.825   -2.35e+05    2.94e+05
Id                        0.9502      1.553      0.612      0.541      -2.097       3.997
MSSubClass              -44.1460     83.026     -0.532      0.595    -207.037     118.745
LotArea                   0.7451      0.108      6.874      0.000       0.532       0.958
OverallQual            6834.4562   1018.778      6.708      0.000    4835.686    8833.227
OverallCond            5791.8768    869.966      6.658      0.000    4085.064    7498.689
YearBuilt               312.6367     75.634      4.134      0.000     164.248     461.026
YearRemodAdd             98.3916     55.284      1.780      0.075     -10.072     206.855
BsmtFinSF1               16.4745      2.894      5.692      0.000      10.796      22.153
BsmtFinSF2                8.5243      5.995      1.422      0.155      -3.237      20.286
BsmtUnfSF                -1.2527      2.776     -0.451      0.652      -6.700       4.195
TotalBsmtSF              23.7461      4.070      5.835      0.000      15.762      31.731
1stFlrSF                 17.6832      6.408      2.760      0.006       5.112      30.255
2ndFlrSF                 37.8252      5.465      6.922      0.000      27.104      48.546
LowQualFinSF            -28.2030     13.968     -2.019      0.044     -55.607      -0.799
GrLivArea                27.3054      5.704      4.787      0.000      16.115      38.496
BsmtFullBath           1133.0294   1980.624      0.572      0.567   -2752.816    5018.875
BsmtHalfBath           -353.8186   3035.753     -0.117      0.907   -6309.752    5602.115
FullBath               3589.3243   2201.641      1.630      0.103    -730.141    7908.789
HalfBath               2100.5290   2103.990      0.998      0.318   -2027.351    6228.409
BedroomAbvGr          -3733.4626   1367.327     -2.730      0.006   -6416.062   -1050.863
KitchenAbvGr          -1.292e+04   5711.650     -2.261      0.024   -2.41e+04   -1710.815
TotRmsAbvGrd           1611.7998    958.825      1.681      0.093    -269.348    3492.947
Fireplaces             6224.4216   2563.472      2.428      0.015    1195.069    1.13e+04
GarageCars             4033.7221   2277.859      1.771      0.077    -435.277    8502.721
GarageArea               18.6372      7.692      2.423      0.016       3.547      33.728
WoodDeckSF               14.8873      5.880      2.532      0.011       3.351      26.424
OpenPorchSF              -3.2280     11.563     -0.279      0.780     -25.914      19.458
EnclosedPorch             2.7081     12.510      0.216      0.829     -21.835      27.251
3SsnPorch                34.0608     22.433      1.518      0.129      -9.951      78.072
ScreenPorch              36.5760     12.538      2.917      0.004      11.978      61.174
PoolArea               

### Model-2

Şimdi de sadece istatiksel olarak anlamlı olan değişkenler ile model olusturalım: ilk modelde istatiksel olarak anlamlı olan bazı değişkenlerin 2.modelde anlamsız olduğu görülüyor.Bunun nedeni, bir değişkeni her eklediğimiz veya çıkarttığımız zaman, model spesifikasyonu değişir. Bu, katsayılar için farklı tahminler ile sonuçlanır. Aynısı, katsayıların istatistiksel anlamlılığı için de geçerlidir.örneğin RoofStyle_Shed değişkeni ilk model için pvalue=0.023 ile anlamlıyken 2.modelde 0.869 değeri ile anlamsız olmustur.Bunun yanında ilk modelde katsayısı 7.737e+04 iken 2.modelde -3453.7212 olmustur.

In [8]:
gecerli_p=list(results.pvalues[results.pvalues < 0.05].index)

In [9]:
Y = house2["SalePrice"]
X = house2[gecerli_p]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.881
Model:                            OLS   Adj. R-squared:                  0.876
Method:                 Least Squares   F-statistic:                     203.8
Date:                Wed, 21 Apr 2021   Prob (F-statistic):               0.00
Time:                        06:58:28   Log-Likelihood:                -16992.
No. Observations:                1460   AIC:                         3.409e+04
Df Residuals:                    1408   BIC:                         3.436e+04
Df Model:                          51                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -9.265e+05   8.51e+04    -10.890      0.000   -1.09e+06    -7.6e+05
LotArea                  0.5684      0.081      6.987      0.000       0.409       0.728
OverallQual           1.185e+04    981.147     12.078      0.000    9926.037    1.38e+04
OverallCond           6732.3706    780.759      8.623      0.000    5200.794    8263.947
YearBuilt              444.3701     42.752     10.394      0.000     360.506     528.235
BsmtFinSF1              19.1072      2.053      9.308      0.000      15.080      23.134
TotalBsmtSF             19.4485      3.455      5.629      0.000      12.671      26.226
1stFlrSF                38.5074      5.135      7.499      0.000      28.435      48.580
2ndFlrSF                32.4379      4.571      7.097      0.000      23.472      41.404
LowQualFinSF           -43.0407     12.660     -3.400      0.001     -67.875     -18.206
GrLivArea               27.9047      4.472      6.239      0.000      19.131      36.678
BedroomAbvGr         -4449.1437   1232.150     -3.611      0.000   -6866.192   -2032.096
KitchenAbvGr         -2.027e+04   3813.190     -5.315      0.000   -2.77e+04   -1.28e+04
Fireplaces            2016.6135   1458.772      1.382      0.167    -844.988    4878.215
GarageArea              50.4726      5.599      9.014      0.000      39.488      61.457
WoodDeckSF               9.2677      6.519      1.422      0.155      -3.521      22.056
ScreenPorch             40.6184     14.132      2.874      0.004      12.896      68.341
PoolArea               637.0411    196.019      3.250      0.001     252.520    1021.562
MoSold                -406.9494    275.415     -1.478      0.140    -947.218     133.319
Neighborhood_Crawfor  2.253e+04   4361.216      5.167      0.000     1.4e+04    3.11e+04
Neighborhood_Edwards -4239.3564   3181.122     -1.333      0.183   -1.05e+04    2000.892
Neighborhood_Mitchel -1.024e+04   4284.720     -2.389      0.017   -1.86e+04   -1831.603
Neighborhood_NAmes   -3056.3945   2327.502     -1.313      0.189   -7622.140    1509.351
Neighborhood_NWAmes  -1.605e+04   3592.496     -4.468      0.000   -2.31e+04   -9004.651
Neighborhood_NoRidge  3.018e+04   4930.992      6.121      0.000    2.05e+04    3.99e+04
Neighborhood_NridgHt  4.367e+04   3775.965     11.565      0.000    3.63e+04    5.11e+04
Neighborhood_StoneBr  4.396e+04   5986.283      7.343      0.000    3.22e+04    5.57e+04
Condition2_PosA       1.936e+04   2.96e+04      0.653      0.514   -3.88e+04    7.75e+04
Condition2_PosN      -1.993e+05   2.05e+04     -9.714      0.000    -2.4e+05   -1.59e+05
RoofStyle_Shed       -3453.7212   2.09e+04     -0.165      0.869   -4.44e+04    3.75e+04
RoofMatl_ClyTile     -5.643e+05   5.19e+04    -10.876      0.000   -6.66e+05   -4.62e+05
RoofMatl_CompShg      1.791e+04   1.22e+04      1.466    

### Model-3

bizim için anlamlı olduğunu düşündüğümüz modelimizi kuralım ve farklılıkları görmek için de etkisini olmadığını düşündüğümüz değişkenlerden de ekleyelim

In [13]:
Y = house2["SalePrice"]
X = house2[['KitchenQual_type','OverallQual','GrLivArea','GarageCars','TotalBsmtSF','YearBuilt',
            "BsmtHalfBath","EnclosedPorch","MiscVal","MSZoning_RH","Street_Grvl",
            "LotShape_IR1","Utilities_NoSeWa","Neighborhood_Blmngtn","Alley_Grvl","Heating_Floor"]]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.784
Model:                            OLS   Adj. R-squared:                  0.782
Method:                 Least Squares   F-statistic:                     328.3
Date:                Wed, 21 Apr 2021   Prob (F-statistic):               0.00
Time:                        07:06:50   Log-Likelihood:                -17424.
No. Observations:                1460   AIC:                         3.488e+04
Df Residuals:                    1443   BIC:                         3.497e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -4.253e+05   9.48e+04     -4.486      0.000   -6.11e+05   -2.39e+05
KitchenQual_type      1.886e+04   2072.269      9.103      0.000    1.48e+04    2.29e+04
OverallQual           1.682e+04   1210.876     13.891      0.000    1.44e+04    1.92e+04
GrLivArea               48.4412      2.502     19.357      0.000      43.532      53.350
GarageCars            1.341e+04   1786.536      7.508      0.000    9908.169    1.69e+04
TotalBsmtSF             28.0652      2.747     10.217      0.000      22.677      33.454
YearBuilt              166.5006     49.678      3.352      0.001      69.052     263.949
BsmtHalfBath          4747.9870   4105.801      1.156      0.248   -3305.991    1.28e+04
EnclosedPorch          -24.2249     17.476     -1.386      0.166     -58.506      10.056
MiscVal                  0.9887      1.965      0.503      0.615      -2.865       4.843
MSZoning_RH          -4005.7106   9415.662     -0.425      0.671   -2.25e+04    1.45e+04
Street_Grvl            270.7964   1.53e+04      0.018      0.986   -2.97e+04    3.02e+04
LotShape_IR1          6535.4457   2143.936      3.048      0.002    2329.881    1.07e+04
Utilities_NoSeWa     -4.692e+04   3.74e+04     -1.256      0.209    -1.2e+05    2.64e+04
Neighborhood_Blmngtn -3.086e+04   9216.888     -3.348      0.001   -4.89e+04   -1.28e+04
Alley_Grvl           -8716.8106   5697.649     -1.530      0.126   -1.99e+04    2459.750
Heating_Floor         3.027e+04   3.73e+04      0.812      0.417   -4.28e+04    1.03e+05
==============================================================================
Omnibus:                      453.423   Durbin-Watson:                   1.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            50990.605
Skew:                          -0.326   Prob(JB):                         0.00
Kurtosis:                      31.944   Cond. No.                     2.67e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.67e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Model-4

Son model için değişken seçimi:

In [16]:
y = house2['SalePrice']
x = house2.drop("SalePrice",axis=1)

from mlxtend.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LinearRegression
feature_select = SequentialFeatureSelector(LinearRegression(),
                                          k_features = 20,
                                          floating = True,
                                          forward = True,
                                          scoring = "r2",
                                          cv = 0)
feature_select.fit(x,y)
list(feature_select.k_feature_names_)

['LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'BsmtFinSF1',
 'TotalBsmtSF',
 'GrLivArea',
 'BedroomAbvGr',
 'KitchenQual_type',
 'Neighborhood_Crawfor',
 'Neighborhood_NoRidge',
 'Neighborhood_NridgHt',
 'Neighborhood_StoneBr',
 'Condition2_PosN',
 'BldgType_1Fam',
 'RoofMatl_ClyTile',
 'ExterQual_Ex',
 'BsmtQual_Ex',
 'BsmtExposure_No',
 'SaleType_New']

model için önemli olduğunu düşündüğümüz değişkenlerle modeli tekrar calıştırırsak R2 değerinin arttığını görüyoruz ve Pvalue değeri 0.05 ten büyük olan değer olmadığı için seçtiğimiz değişkenler model için anlamlıdır.Ev fiyatları üzerinde belirgin etkileri vardır.

In [14]:
Y = house2["SalePrice"]
X = house2[list(feature_select.k_feature_names_)]

X = sm.add_constant(X)
results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.895
Model:                            OLS   Adj. R-squared:                  0.894
Method:                 Least Squares   F-statistic:                     616.5
Date:                Wed, 21 Apr 2021   Prob (F-statistic):               0.00
Time:                        07:08:19   Log-Likelihood:                -16895.
No. Observations:                1460   AIC:                         3.383e+04
Df Residuals:                    1439   BIC:                         3.394e+04
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                -9.472e+05   6.66e+04    -14.215      0.000   -1.08e+06   -8.17e+05
LotArea                  0.6013      0.074      8.110      0.000       0.456       0.747
OverallQual           1.029e+04    886.462     11.603      0.000    8546.824     1.2e+04
OverallCond           5899.5565    704.257      8.377      0.000    4518.077    7281.036
YearBuilt              447.3767     34.206     13.079      0.000     380.278     514.476
BsmtFinSF1              19.4217      1.896     10.244      0.000      15.703      23.141
TotalBsmtSF             19.0317      2.262      8.413      0.000      14.594      23.469
GrLivArea               63.1448      2.273     27.783      0.000      58.686      67.603
BedroomAbvGr         -6198.4365   1098.935     -5.640      0.000   -8354.122   -4042.751
KitchenQual_type      7819.1638   1513.059      5.168      0.000    4851.127    1.08e+04
Neighborhood_Crawfor  2.328e+04   3859.732      6.032      0.000    1.57e+04    3.09e+04
Neighborhood_NoRidge  4.486e+04   4502.044      9.965      0.000     3.6e+04    5.37e+04
Neighborhood_NridgHt  2.535e+04   3714.199      6.826      0.000    1.81e+04    3.26e+04
Neighborhood_StoneBr  3.996e+04   5539.714      7.214      0.000    2.91e+04    5.08e+04
Condition2_PosN      -2.452e+05    1.9e+04    -12.908      0.000   -2.82e+05   -2.08e+05
BldgType_1Fam          1.95e+04   1952.644      9.987      0.000    1.57e+04    2.33e+04
RoofMatl_ClyTile     -6.579e+05   2.88e+04    -22.849      0.000   -7.14e+05   -6.01e+05
ExterQual_Ex          3.612e+04   4632.360      7.797      0.000     2.7e+04    4.52e+04
BsmtQual_Ex           2.751e+04   3338.124      8.241      0.000     2.1e+04    3.41e+04
BsmtExposure_No      -1.023e+04   1524.948     -6.711      0.000   -1.32e+04   -7242.759
SaleType_New           1.81e+04   2861.150      6.328      0.000    1.25e+04    2.37e+04
==============================================================================
Omnibus:                      415.904   Durbin-Watson:                   1.978
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8431.662
Skew:                           0.812   Prob(JB):                         0.00
Kurtosis:                      14.660   Cond. No.                     1.44e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.44e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""